gremlin-go/driver/gorillaTransporter.go

```diff
@@ -68,16 +73,47 @@ func (transporter *gorillaTransporter) Write(data []byte) (err error) {
 	return err
 }
 
-func (transporter *gorillaTransporter) Read() (bytes []byte, err error) {
+func (transporter *gorillaTransporter) Read() ([]byte, error) {
 	if transporter.connection == nil {
-		err = transporter.Connect()
+		err := transporter.Connect()
 		if err != nil {
-			return
+			return nil, err
+		}
+	}
+	transporter.connection.SetPongHandler(func(string) error {
+		err := transporter.connection.SetReadDeadline(time.Now().Add(time.Second))
+		if err != nil {
+			return err
 		}
+		return nil
+	})
+
+	err := transporter.connection.SetReadDeadline(time.Now().Add(time.Second))
+	if err != nil {
+		return nil, err
 	}
 
-	_, bytes, err = transporter.connection.ReadMessage()
-	return
+	failureCount := 0
+	for {
+		_, bytes, err := transporter.connection.ReadMessage()
+		if err == nil {
+			return bytes, nil
+		}
+		failureCount += 1
+		if failureCount > maxFailCount {
+			return nil, errors.New(fmt.Sprintf("failed to read from socket more than %d times", maxFailCount))
+		}
+
+		// Try pinging server.
+		err = transporter.connection.SetWriteDeadline(time.Now().Add(500 * time.Millisecond))
```

Comment:
In general I'm not sure if a static timeout for something that can be out of a user's control to be a good idea, but this is just to bring it to attention.

Reply:
I will make a ticket to make it configurable.